File: src/dojopool/schemas/tournament.py
"""Schemas for tournament-related data validation."""

from marshmallow import Schema, ValidationError, fields, validate, validates


class TournamentSchema(Schema):
    """Schema for tournament data."""

    id = fields.Int(dump_only=True)
    name = fields.Str(required=True, validate=validate.Length(min=1, max=100))
    description = fields.Str(validate=validate.Length(max=500))
    venue_id = fields.Int(required=True)
    format = fields.Str(
        required=True,
        validate=validate.OneOf(
            ["single_elimination", "double_elimination", "round_robin", "swiss"]
        ),
    )
    status = fields.Str(dump_only=True)
    start_date = fields.DateTime(required=True)
    end_date = fields.DateTime(required=True)
    registration_deadline = fields.DateTime()
    max_participants = fields.Int(validate=validate.Range(min=2))
    entry_fee = fields.Float(validate=validate.Range(min=0))
    prize_pool = fields.Float(validate=validate.Range(min=0))
    rules = fields.Str(validate=validate.Length(max=1000))
    bracket_data = fields.Dict(dump_only=True)
    created_at = fields.DateTime(dump_only=True)
    updated_at = fields.DateTime(dump_only=True)

    @validates("end_date")
    def validate_end_date(self, value):
        """Validate end date is after start date."""
        if (
            value
            and self.context.get("start_date")
            and value <= self.context["start_date"]
        ):
            raise ValidationError("End date must be after start date")

    @validates("registration_deadline")
    def validate_registration_deadline(self, value):
        """Validate registration deadline is before start date."""
        if (
            value
            and self.context.get("start_date")
            and value >= self.context["start_date"]
        ):
            raise ValidationError("Registration deadline must be before start date")


class TournamentParticipantSchema(Schema):
    """Schema for tournament participant data."""

    id = fields.Int(dump_only=True)
    tournament_id = fields.Int(dump_only=True)
    user_id = fields.Int(dump_only=True)
    seed = fields.Int(dump_only=True)
    registered_at = fields.DateTime(dump_only=True)
    checked_in = fields.Bool(dump_only=True)
    checked_in_at = fields.DateTime(dump_only=True)
    eliminated = fields.Bool(dump_only=True)
    final_rank = fields.Int(dump_only=True)
    prize_amount = fields.Float(dump_only=True)
    stats = fields.Dict(dump_only=True)
    user = fields.Nested(
        "UserSchema", only=("id", "username", "avatar_url"), dump_only=True
    )


class TournamentMatchSchema(Schema):
    """Schema for tournament match data."""

    id = fields.Int(dump_only=True)
    tournament_id = fields.Int(dump_only=True)
    round_number = fields.Int(dump_only=True)
    match_number = fields.Int(dump_only=True)
    player1_id = fields.Int(dump_only=True)
    player2_id = fields.Int(dump_only=True)
    winner_id = fields.Int()
    next_match_id = fields.Int(dump_only=True)
    status = fields.Str(dump_only=True)
    start_time = fields.DateTime()
    end_time = fields.DateTime()
    table_number = fields.Int(validate=validate.Range(min=1))
    score = fields.Dict()
    stats = fields.Dict()
    notes = fields.Str(validate=validate.Length(max=500))
    player1 = fields.Nested(TournamentParticipantSchema, dump_only=True)
    player2 = fields.Nested(TournamentParticipantSchema, dump_only=True)
    winner = fields.Nested(TournamentParticipantSchema, dump_only=True)

    @validates("end_time")
    def validate_end_time(self, value):
        """Validate end time is after start time."""
        if (
            value
            and self.context.get("start_time")
            and value <= self.context["start_time"]
        ):
            raise ValidationError("End time must be after start time")

File: src/types/tournament.ts
/**
 * Tournament Type Definitions
 * 
 * Comprehensive type definitions for the unified tournament system
 */

export interface Tournament {
  id: number;
  name: string;
  description?: string;
  venueId: number;
  organizerId: number;
  startDate: Date;
  endDate: Date;
  registrationDeadline: Date;
  maxParticipants: number;
  entryFee: number;
  prizePool: number;
  status: TournamentStatus;
  format: TournamentFormat;
  rules?: string;
  createdAt: Date;
  updatedAt: Date;
  participantCount: number;
  venue?: Venue;
  organizer?: User;
}

export interface TournamentParticipant {
  id: number;
  tournamentId: number;
  userId: number;
  username: string;
  seed?: number;
  status: 'registered' | 'checked_in' | 'eliminated' | 'withdrawn';
  registrationDate: Date;
  checkInTime?: Date;
  eliminationRound?: number;
  finalPlacement?: number;
  user?: User;
}

export interface TournamentMatch {
  id: string;
  tournamentId: number;
  roundNumber: number;
  matchNumber: number;
  player1Id?: number;
  player2Id?: number;
  player1Name?: string;
  player2Name?: string;
  player1Score?: number;
  player2Score?: number;
  winnerId?: number;
  status: MatchStatus;
  bracketType: 'winners' | 'losers' | 'consolation';
  isBye?: boolean;
  isForfeit?: boolean;
  startTime?: Date;
  endTime?: Date;
  nextMatchId?: string;
  previousMatchIds?: string[];
}

export interface TournamentRound {
  id: number;
  tournamentId: number;
  roundNumber: number;
  name: string;
  matches: TournamentMatch[];
  isActive: boolean;
  isCompleted: boolean;
  startTime?: Date;
  endTime?: Date;
}

export interface TournamentBracket {
  id: string;
  tournamentId: number;
  type: TournamentFormat;
  status: TournamentStatus;
  rounds: TournamentRound[];
  nodes: BracketNode[];
  participants: TournamentParticipant[];
  currentRound: number;
  totalRounds: number;
  totalMatches: number;
  completedMatches: number;
  startDate: Date;
  endDate?: Date;
  seedingMethod: SeedingMethod;
  consolationRounds: boolean;
}

export interface BracketNode {
  id: string;
  matchId?: number;
  player1Id?: number;
  player2Id?: number;
  player1Name?: string;
  player2Name?: string;
  player1Score?: number;
  player2Score?: number;
  winnerId?: number;
  status: MatchStatus;
  round: number;
  matchNumber: number;
  bracketType: 'winners' | 'losers' | 'consolation';
  nextMatchId?: string;
  previousMatchIds?: string[];
  isBye?: boolean;
  isForfeit?: boolean;
}

export interface TournamentConfig {
  id?: number;
  name: string;
  description?: string;
  venueId: number;
  organizerId: number;
  startDate: Date;
  endDate: Date;
  registrationDeadline: Date;
  maxParticipants: number;
  entryFee: number;
  prizePool: number;
  format: TournamentFormat;
  rules?: string;
  seedingMethod?: SeedingMethod;
  autoStart?: boolean;
  checkInRequired?: boolean;
  consolationRounds?: boolean;
}

export const TournamentFormat = {
  SINGLE_ELIMINATION: 'single_elimination',
  DOUBLE_ELIMINATION: 'double_elimination',
  ROUND_ROBIN: 'round_robin',
  SWISS: 'swiss',
  CONSOLATION: 'consolation'
} as const;

export type TournamentFormat = typeof TournamentFormat[keyof typeof TournamentFormat];

export const TournamentStatus = {
  DRAFT: 'draft',
  REGISTRATION: 'registration',
  CHECK_IN: 'check_in',
  IN_PROGRESS: 'in_progress',
  COMPLETED: 'completed',
  CANCELLED: 'cancelled'
} as const;

export type TournamentStatus = typeof TournamentStatus[keyof typeof TournamentStatus];

export const SeedingMethod = {
  RANDOM: 'random',
  RATING: 'rating',
  RANKING: 'ranking',
  MANUAL: 'manual',
  TOURNAMENT_HISTORY: 'tournament_history'
} as const;

export type SeedingMethod = typeof SeedingMethod[keyof typeof SeedingMethod];

export const MatchStatus = {
  PENDING: 'pending',
  IN_PROGRESS: 'in_progress',
  COMPLETED: 'completed',
  BYE: 'bye',
  FORFEIT: 'forfeit'
} as const;

export type MatchStatus = typeof MatchStatus[keyof typeof MatchStatus];

// Import types that are referenced
export interface User {
  id: number;
  username: string;
  email: string;
  avatarUrl?: string;
  rating?: number;
  ranking?: number;
}

export interface Venue {
  id: number;
  name: string;
  address: string;
  city: string;
  state: string;
  country: string;
  latitude?: number;
  longitude?: number;
}
